### PR TITLE
Add Room database with UseCases

### PR DESCRIPTION
**Pull request by Gowsky**
Add Room database with UseCases related to loading and inserting measurements.

### CODE REVIEW

File: domain/build.gradle [hunk 1]
```diff
@@ -4,6 +4,8 @@ plugins {
     id "kotlin-kapt"
 }
 
+apply plugin: 'kotlinx-serialization'
```
Comment (ghost) [inline domain/build.gradle]:
```suggestion
    id "kotlin-kapt"
    id "kotlinx-serialization"
}
```

File: domain/src/main/kotlin/com/patrykandpatryk/liftapp/domain/mapper/Mapper.kt [hunk 2]
```diff
@@ -5,4 +5,6 @@ interface Mapper<I, O> {
     fun map(input: I): O
 
     operator fun invoke(input: I): O = map(input)
+
+    operator fun invoke(transform: Iterable<I>): List<O> = transform.map { input -> map(input) }
```
Comment (ghost) [inline domain/src/main/kotlin/com/patrykandpatryk/liftapp/domain/mapper/Mapper.kt]:
```suggestion
    operator fun invoke(transform: Iterable<I>): List<O> = transform.map(::map)
```

File: domain/src/main/kotlin/com/patrykandpatryk/liftapp/domain/measurement/InsertMeasurementEntryUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package com.patrykandpatryk.liftapp.domain.measurement
+
+import java.util.*
```
Comment (ghost) [inline domain/src/main/kotlin/com/patrykandpatryk/liftapp/domain/measurement/InsertMeasurementEntryUseCase.kt]:
```suggestion
import java.util.Date
```

File: domain/src/main/kotlin/com/patrykandpatryk/liftapp/domain/measurement/MeasurementRepository.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package com.patrykandpatryk.liftapp.domain.measurement
+
+import com.patrykandpatryk.liftapp.domain.model.Name
+import java.util.*
```
Comment (ghost) [inline domain/src/main/kotlin/com/patrykandpatryk/liftapp/domain/measurement/MeasurementRepository.kt]:
```suggestion
import java.util.Date
```

File: functionality/database/build.gradle [hunk 5]
```diff
@@ -1,4 +1,9 @@
 apply from: "$rootDir/gradle/functionality-module-base.gradle"
+apply plugin: 'kotlinx-serialization'
```
Comment (ghost) [inline functionality/database/build.gradle]:
```suggestion
apply plugin: "kotlinx-serialization"
```

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/Database.kt [hunk 6]
```diff
@@ -0,0 +1,36 @@
+package pl.patrykgoworowski.mintlift.functionality.database
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/Database.kt]:
```suggestion
package com.patrykandpatryk.liftapp.functionality.database
```

I’d also suggest renaming the packages accordingly.

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/Database.kt [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+import androidx.room.BuiltInTypeConverters
+import androidx.room.RoomDatabase
+import androidx.room.TypeConverters
+import pl.patrykgoworowski.mintlift.functionality.database.converter.DateConverters
+import pl.patrykgoworowski.mintlift.functionality.database.converter.JsonConverters
+import pl.patrykgoworowski.mintlift.functionality.database.measurement.MeasurementEntity
+import pl.patrykgoworowski.mintlift.functionality.database.measurement.MeasurementDao
+import pl.patrykgoworowski.mintlift.functionality.database.measurement.MeasurementEntryEntity
+import pl.patrykgoworowski.mintlift.functionality.database.measurement.MeasurementWithLatestEntryView
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/Database.kt]:
```suggestion
import com.patrykandpatryk.liftapp.functionality.database.converter.DateConverters
import com.patrykandpatryk.liftapp.functionality.database.converter.JsonConverters
import com.patrykandpatryk.liftapp.functionality.database.measurement.MeasurementEntity
import com.patrykandpatryk.liftapp.functionality.database.measurement.MeasurementDao
import com.patrykandpatryk.liftapp.functionality.database.measurement.MeasurementEntryEntity
import com.patrykandpatryk.liftapp.functionality.database.measurement.MeasurementWithLatestEntryView
```

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/converter/DateConverters.kt [hunk 8]
```diff
@@ -0,0 +1,20 @@
+package pl.patrykgoworowski.mintlift.functionality.database.converter
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/converter/DateConverters.kt]:
```suggestion
package com.patrykandpatryk.liftapp.functionality.database.converter
```

I’d also suggest renaming the packages accordingly.

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/converter/DateConverters.kt [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+import androidx.room.ProvidedTypeConverter
+import androidx.room.TypeConverter
+import java.text.SimpleDateFormat
+import java.util.Date
+import javax.inject.Inject
+import pl.patrykgoworowski.mintlift.functionality.database.di.DatabaseDateFormat
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/converter/DateConverters.kt]:
```suggestion
import com.patrykandpatryk.liftapp.functionality.database.di.DatabaseDateFormat
```

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/converter/JsonConverters.kt [hunk 10]
```diff
@@ -0,0 +1,28 @@
+package pl.patrykgoworowski.mintlift.functionality.database.converter
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/converter/JsonConverters.kt]:
```suggestion
package com.patrykandpatryk.liftapp.functionality.database.converter
```

I’d also suggest renaming the packages accordingly.

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/di/DatabaseDateFormat.kt [hunk 11]
```diff
@@ -0,0 +1,6 @@
+package pl.patrykgoworowski.mintlift.functionality.database.di
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/di/DatabaseDateFormat.kt]:
```suggestion
package com.patrykandpatryk.liftapp.functionality.database.di
```

I’d also suggest renaming the packages accordingly.

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/di/DatabaseModule.kt [hunk 12]
```diff
@@ -0,0 +1,70 @@
+package pl.patrykgoworowski.mintlift.functionality.database.di
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/di/DatabaseModule.kt]:
```suggestion
package com.patrykandpatryk.liftapp.functionality.database.di
```

I’d also suggest renaming the packages accordingly.

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/di/DatabaseModule.kt [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.*
+import javax.inject.Singleton
+import pl.patrykgoworowski.mintlift.functionality.database.Database
+import pl.patrykgoworowski.mintlift.functionality.database.converter.DateConverters
+import pl.patrykgoworowski.mintlift.functionality.database.converter.JsonConverters
+import pl.patrykgoworowski.mintlift.functionality.database.measurement.MeasurementDao
+import pl.patrykgoworowski.mintlift.functionality.database.measurement.MeasurementEntryEntityToDomainMapper
+import pl.patrykgoworowski.mintlift.functionality.database.measurement.MeasurementRepositoryImpl
+import pl.patrykgoworowski.mintlift.functionality.database.measurement.MeasurementWithLatestEntryToDomainMeasurementMapper
+import pl.patrykgoworowski.mintlift.functionality.database.measurement.MeasurementWithLatestEntryView
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/di/DatabaseModule.kt]:
```suggestion
import com.patrykandpatryk.liftapp.functionality.database.Database
import com.patrykandpatryk.liftapp.functionality.database.converter.DateConverters
import com.patrykandpatryk.liftapp.functionality.database.converter.JsonConverters
import com.patrykandpatryk.liftapp.functionality.database.measurement.MeasurementDao
import com.patrykandpatryk.liftapp.functionality.database.measurement.MeasurementEntryEntityToDomainMapper
import com.patrykandpatryk.liftapp.functionality.database.measurement.MeasurementRepositoryImpl
import com.patrykandpatryk.liftapp.functionality.database.measurement.MeasurementWithLatestEntryToDomainMeasurementMapper
import com.patrykandpatryk.liftapp.functionality.database.measurement.MeasurementWithLatestEntryView
```

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementDao.kt [hunk 14]
```diff
@@ -0,0 +1,22 @@
+package pl.patrykgoworowski.mintlift.functionality.database.measurement
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementDao.kt]:
```suggestion
package com.patrykandpatryk.liftapp.functionality.database.measurement
```

I’d also suggest renaming the packages accordingly.

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementEntity.kt [hunk 15]
```diff
@@ -0,0 +1,16 @@
+package pl.patrykgoworowski.mintlift.functionality.database.measurement
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementEntity.kt]:
```suggestion
package com.patrykandpatryk.liftapp.functionality.database.measurement
```

I’d also suggest renaming the packages accordingly.

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementEntryEntity.kt [hunk 16]
```diff
@@ -0,0 +1,37 @@
+package pl.patrykgoworowski.mintlift.functionality.database.measurement
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementEntryEntity.kt]:
```suggestion
package com.patrykandpatryk.liftapp.functionality.database.measurement
```

I’d also suggest renaming the packages accordingly.

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementMappers.kt [hunk 17]
```diff
@@ -0,0 +1,32 @@
+package pl.patrykgoworowski.mintlift.functionality.database.measurement
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementMappers.kt]:
```suggestion
package com.patrykandpatryk.liftapp.functionality.database.measurement
```

I’d also suggest renaming the packages accordingly.

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementRepositoryImpl.kt [hunk 18]
```diff
@@ -0,0 +1,55 @@
+package pl.patrykgoworowski.mintlift.functionality.database.measurement
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementRepositoryImpl.kt]:
```suggestion
package com.patrykandpatryk.liftapp.functionality.database.measurement
```

I’d also suggest renaming the packages accordingly.

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementWithLatestEntryView.kt [hunk 19]
```diff
@@ -0,0 +1,19 @@
+@file:Suppress("Indentation")
+package pl.patrykgoworowski.mintlift.functionality.database.measurement
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementWithLatestEntryView.kt]:
```suggestion
package com.patrykandpatryk.liftapp.functionality.database.measurement
```

I’d also suggest renaming the packages accordingly.

**Review by ghost — CHANGES_REQUESTED** [review]

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/di/DatabaseModule.kt [hunk 20]
```diff
@@ -0,0 +1,70 @@
+package pl.patrykgoworowski.mintlift.functionality.database.di
+
+import android.app.Application
+import androidx.room.Room
+import com.patrykandpatryk.liftapp.domain.Constants
+import com.patrykandpatryk.liftapp.domain.mapper.Mapper
+import com.patrykandpatryk.liftapp.domain.measurement.Measurement
```
Comment (ghost) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/di/DatabaseModule.kt]:
```suggestion
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementWithLatestEntryView.kt [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+import androidx.room.DatabaseView
+import androidx.room.Embedded
+
+@DatabaseView(
+    value = "SELECT measurement.*, latest_entry.* FROM measurement " +
+            "LEFT JOIN " +
+            "(SELECT * FROM measurement_entry as entry WHERE entry.timestamp in" +
+            "(SELECT MAX(E.timestamp) FROM measurement_entry as E GROUP BY E.parent_id) " +
+            "ORDER BY entry.entry_id DESC) " +
+            "as latest_entry on measurement.id = latest_entry.parent_id",
```
Comment (patrykmichalik) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementWithLatestEntryView.kt:13]:
```suggestion
    value = "SELECT measurement.*, latest_entry.* FROM measurement " +
            "LEFT JOIN " +
            "(SELECT * FROM measurement_entry AS entry WHERE entry.timestamp IN" +
            "(SELECT MAX(E.timestamp) FROM measurement_entry AS E GROUP BY E.parent_id) " +
            "ORDER BY entry.entry_id DESC) " +
            "AS latest_entry ON measurement.id = latest_entry.parent_id",
```

**Review by patrykmichalik — CHANGES_REQUESTED** [review]

File: functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementEntryEntity.kt [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+)
+class MeasurementEntryEntity(
+    @PrimaryKey(autoGenerate = true)
+    @ColumnInfo(name = "entry_id")
+    val id: Long = 0,
+    @ColumnInfo(name = "parent_id")
+    val parentId: Long,
+    val values: MeasurementValues,
+    val timestamp: Date,
+)
```
Comment (patrykmichalik) [inline functionality/database/src/main/kotlin/pl/patrykgoworowski/mintlift/functionality/database/measurement/MeasurementEntryEntity.kt:37]:
```suggestion
)
class MeasurementEntryEntity(
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "entry_id", index = true)
    val id: Long = 0,
    @ColumnInfo(name = "parent_id", index = true)
    val parentId: Long,
    @ColumnInfo(index = true)
    val values: MeasurementValues,
    val timestamp: Date,
)
```

**Review by patrykmichalik — CHANGES_REQUESTED** [review]

File: functionality/database/src/main/kotlin/com/patrykandpatryk/liftapp/functionality/database/measurement/MeasurementWithLatestEntryView.kt [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+import androidx.room.DatabaseView
+import androidx.room.Embedded
+
+@DatabaseView(
+    value = "SELECT measurement.*, latest_entry.* FROM measurement " +
+            "LEFT JOIN " +
+            "(SELECT * FROM measurement_entry as entry WHERE entry.timestamp in" +
+            "(SELECT MAX(E.timestamp) FROM measurement_entry as E GROUP BY E.parent_id) " +
+            "ORDER BY entry.entry_id DESC) " +
+            "as latest_entry on measurement.id = latest_entry.parent_id",
```
Comment (ghost) [inline functionality/database/src/main/kotlin/com/patrykandpatryk/liftapp/functionality/database/measurement/MeasurementWithLatestEntryView.kt]:
```suggestion
    value = "SELECT measurement.*, latest_entry.* FROM measurement " +
            "LEFT JOIN " +
            "(SELECT * FROM measurement_entry AS entry WHERE entry.timestamp IN" +
            "(SELECT MAX(E.timestamp) FROM measurement_entry AS E GROUP BY E.parent_id) " +
            "ORDER BY entry.entry_id DESC) " +
            "AS latest_entry ON measurement.id = latest_entry.parent_id",
```

File: functionality/database/src/main/kotlin/com/patrykandpatryk/liftapp/functionality/database/measurement/MeasurementEntryEntity.kt [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+)
+class MeasurementEntryEntity(
+    @PrimaryKey(autoGenerate = true)
+    @ColumnInfo(name = "entry_id")
+    val id: Long = 0,
+    @ColumnInfo(name = "parent_id")
+    val parentId: Long,
+    val values: MeasurementValues,
+    val timestamp: Date,
+)
```
Comment (ghost) [inline functionality/database/src/main/kotlin/com/patrykandpatryk/liftapp/functionality/database/measurement/MeasurementEntryEntity.kt]:
```suggestion
)
class MeasurementEntryEntity(
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "entry_id", index = true)
    val id: Long = 0,
    @ColumnInfo(name = "parent_id", index = true)
    val parentId: Long,
    val values: MeasurementValues,
    @ColumnInfo(index = true)
    val timestamp: Date,
)
```

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**